### PR TITLE
Re-structuring garbage collector and adding exception and match.

### PR DESCRIPTION
**Pull request by mohammedzee1000**
This allows user to specify exceptions as well as matches as regex.
It allows allows user to completely skip out on providing an index.

Signed-off-by: Mohammed Zeeshan Ahmed <mohammed.zee1000@gmail.com>

### CODE REVIEW

File: container_pipeline/cleanup_registry/__init__.py [hunk 1]
```diff
@@ -24,24 +24,36 @@ def get_args_parser():
         help="Set if the registry talks over https",
         action="store_true"
     )
+    parser.add_argument(
+        "--indexlocation",
+        help="Specify location where you want to either clone index, or where you have already cloned it. Defaults to "
```
Comment (navidshaikh) [inline container_pipeline/cleanup_registry/__init__.py]: lines exceeding 80 chars

File: container_pipeline/cleanup_registry/__init__.py [hunk 2]
```diff
@@ -24,24 +24,36 @@ def get_args_parser():
...
+             "./c_i",
+        default="./c_i"
+    )
     parser.add_argument(
         "-l",
         "--localindex",
-        help="Set if you want to use locally available index. Note the index.d should be in ./c_i directory",
+        help="Set if you want to use locally available index. Note the index.d should be in index location",
         action="store_true"
```
Comment (navidshaikh) [inline container_pipeline/cleanup_registry/__init__.py]: lines exceeding 80 chars

File: container_pipeline/cleanup_registry/__init__.py [hunk 3]
```diff
@@ -24,24 +24,36 @@ def get_args_parser():
...
     )
     parser.add_argument(
         "-i",
-        "--indexurl",
-        help="The git url of index to be cloned.",
-        default="https://github.com/centos/container-index"
+        "--indexgiturl",
+        help="Specify the index git url to clone. If it is not specified, then no repo will be cloned.",
```
Comment (navidshaikh) [inline container_pipeline/cleanup_registry/__init__.py]: lines exceeding 80 chars

File: container_pipeline/cleanup_registry/garbagecollector.py [hunk 4]
```diff
@@ -10,53 +12,118 @@ class GarbageCollector(object):
     """
 
     def __init__(self, registry_host="127.0.0.1", registry_port="5000", registry_secure=False, local_index=False,
-                 index_git="https://github.com/centos/container-index", verbose=True, collect=True):
+                 index_git=None, index_location="./c_i", verbose=True, collect=True):
```
Comment (navidshaikh) [inline container_pipeline/cleanup_registry/garbagecollector.py]: I think default is already added at https://github.com/CentOS/container-pipeline-service/pull/416/files#diff-37f1187e22ecd347ea22a51bc9ba767bR31 for `index_location`, why to repeat?

File: container_pipeline/cleanup_registry/garbagecollector.py [hunk 5]
```diff
@@ -10,53 +12,118 @@ class GarbageCollector(object):
...
+        """
+        Query a cloned or local container index.
+        :return:
+        """
+        index_files = glob(self._index_d_location + "/*.yml")
+        for index_file in index_files:
+            if "index_template" not in index_file:
+                data = lib.load_yaml(index_file)
+                if "Projects" not in data:
+                    raise Exception("Invalid index file")
```
Comment (navidshaikh) [inline container_pipeline/cleanup_registry/garbagecollector.py]: should this log issue for file in process and continue processing rest of the files in index ?

File: container_pipeline/cleanup_registry/garbagecollector.py [hunk 6]
```diff
@@ -10,53 +12,118 @@ class GarbageCollector(object):
...
             print("Gathering registry metadata...")
         self._registry_info = lib.RegistryInfo(self._registry_url)
-        self.index_containers = {}
-        self.mismatched = {}
 
-    def _delete_mismatched(self):
+    def _query_index(self):
+        """
+        Query a cloned or local container index.
+        :return:
```
Comment (navidshaikh) [inline container_pipeline/cleanup_registry/garbagecollector.py]: empty return?

File: container_pipeline/cleanup_registry/garbagecollector.py [hunk 7]
```diff
@@ -10,53 +12,118 @@ class GarbageCollector(object):
...
+                    if container_name not in self._index_containers:
+                        #
+                        self._index_containers[container_name] = []
+                        if not BuildTracker(container_tag).is_running():
+                            self._index_containers[container_name].append(desired_tag)
+
+    def _prep_index(self):
+        """
+        Prepares the container index, and queries the same.
+        :return:
```
Comment (navidshaikh) [inline container_pipeline/cleanup_registry/garbagecollector.py]: empty return?

File: container_pipeline/cleanup_registry/garbagecollector.py [hunk 8]
```diff
@@ -10,53 +12,118 @@ class GarbageCollector(object):
...
+        index_preped = False
+        if self._local_index:
+            # If local index is set, check if index files are present at expected location.
+            if not lib.path.exists(self._index_location):
+                raise Exception("Local index specified, but does not exist")
+            index_preped = True
+        elif self._index_git:
+            # Otherwise check if index git is specified, if so clone it.
+            if self._verbose:
+                print("Cloning container index...")
```
Comment (navidshaikh) [inline container_pipeline/cleanup_registry/garbagecollector.py]: a better implementation of print based on verbose flag could be

```
def _print(msg, verbose=True):
   if verbose:
           print msg
```
and have all the print statements replaced via `_print`
where in, for mandatory print occurrences you can just
```
_print("I must print on stdout.")
```
or for optional (if verbose), you can
```
_print("I could print if you asked.", versbose=verbose_falg_value)
```

File: container_pipeline/cleanup_registry/garbagecollector.py [hunk 9]
```diff
@@ -10,53 +12,118 @@ class GarbageCollector(object):
...
+            index_preped = True
+
+        if index_preped:
+            # If there is an index to query, query it and populate image list
+            self._query_index()
+
+    def _prep_lists(self):
+        """
+        Prepares the exceptions and matches list
+        :return:
```
Comment (navidshaikh) [inline container_pipeline/cleanup_registry/garbagecollector.py]: empty return

File: container_pipeline/cleanup_registry/garbagecollector.py [hunk 10]
```diff
@@ -73,46 +140,60 @@ def _delete_mismatched(self):
...
+                        for exp in self._gc_exceptions:
+                            if exp.match(image_name):
+                                exception_list_found = True
+                                break
+
+                    if match_only_check:
+                        for exp in self._gc_match_only:
+                            if exp.match(image_name):
+                                match_only_list_found = True
+                                break
```
Comment (navidshaikh) [inline container_pipeline/cleanup_registry/garbagecollector.py]: I recommend you add comments to operations performed in this block.

File: container_pipeline/cleanup_registry/garbagecollector.py (hunk 5 above)
Comment (mohammedzee1000) [reply container_pipeline/cleanup_registry/garbagecollector.py]: @navidshaikh no, if an index file is wrong and we allow it to proceed, we might end up with incorrect images being deleted.

**Comment by bamachrn** [comment]
LGTM, merging it. lets make sure this is 100% tested in pre-prod

**Pull request merged**